File: barbieland/submissions/accepted/barbieland_triangles.py
import sys

sys.setrecursionlimit(10 ** 9)


def solve(N: int, M: int, Q: int, U: list[int], V: list[int], W: list[int], A: list[int], B: list[int]):
    nodes_xor = [0 for _ in range(N + 1)]
    graph = [[] for i in range(N + 1)]
    for i in range(M):
        graph[U[i]].append((V[i], W[i]))
        graph[V[i]].append((U[i], W[i]))

    dfs_tree = [[] for i in range(N + 1)]
    visited = set()
    basis = []

    def add_to_basis(x):
        for y in basis:
            x = min(x, y ^ x)
        if x != 0:
            basis.append(x)

    def dfs(u):
        visited.add(u)
        for v, w in graph[u]:
            if v not in visited:
                nodes_xor[v] = nodes_xor[u] ^ w
                dfs_tree[u].append(v)
                dfs(v)

    def triangles(u):
        for v in dfs_tree[u]:
            triangles(v)
        for v, w in graph[u]:
            add_to_basis(w ^ nodes_xor[u] ^ nodes_xor[v])

    dfs(1)
    triangles(1)

    for i in range(Q):
        answer = nodes_xor[A[i]] ^ nodes_xor[B[i]]
        for b in basis:
            answer = max(answer, answer ^ b)
        print(answer)


def main():
    T = int(input())
    for _ in range(T):
        N, M, Q = [int(i) for i in input().split()]
        U, V, W = map(list, zip(*(map(int, input().split()) for _ in range(M))))
        A, B = map(list, zip(*(map(int, input().split()) for _ in range(Q))))
        solve(N, M, Q, U, V, W, A, B)


if __name__ == '__main__':
    main()

File: barbieland/submissions/accepted/barbieland_triangles.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;

/**
 * Output Q lines, where the i-th line contains the maximum
 * fuel charge possible for the i-th errand
 *     
 * N: the number of universes
 * M: the number of portals
 * Q: the number of queries
 * U: the list of U_i for each portal
 * V: the list of V_i for each portal
 * W: the list of W_i for each portal
 * A: the list of A_i for each errand
 * B: the list of B_i for each errand
 * 
 */
void solve(int N, int M, int Q, vector<int>& U, vector<int>& V, vector<ll>& W, vector<int>& A, vector<int>& B) {

    vector<ll> nodes_xor(N + 1);
    vector<vector<pair<int,ll>>> g(N + 1);
    for (int i = 0; i < M; ++i) {
        g[U[i]].emplace_back(V[i], W[i]);
        g[V[i]].emplace_back(U[i], W[i]);
    }

    vector<vector<int>> dfs_tree(N + 1);
    vector<bool> visited(N + 1, false);
    vector<ll> basis;

    function<void(ll)> add_to_basis = [&](ll x) {
        for (ll y : basis) x = min(x, y ^ x);
        if (x) basis.push_back(x);
    };

    // Create DFS tree
    function<void(int)> dfs = [&](int u) {
        visited[u] = true;
        for (auto& [v, w] : g[u]) {
            if (!visited[v]) {
                nodes_xor[v] = nodes_xor[u] ^ w;
                dfs_tree[u].push_back(v);
                dfs(v);
            }
        }
    };

    function<void(int)> triangles = [&](int u) {
        for (int v : dfs_tree[u]) triangles(v);
        for (auto& [v, w] : g[u]) {
            // Triange 1 -> u -> v
            add_to_basis(w ^ nodes_xor[u] ^ nodes_xor[v]);
        }
    };

    dfs(1);
    triangles(1);

    for (int i = 0; i < Q; ++i) {
        ll answer = nodes_xor[A[i]] ^ nodes_xor[B[i]];
        for (ll b : basis) answer = max(answer, answer ^ b);
        cout << answer << '\n';
    }

}

int main() {
    int T;
    cin >> T;
    for (int i = 0; i < T; i++) {
        int N, M, Q;
        cin >> N >> M >> Q;
        vector<int> U(M), V(M), A(Q), B(Q);
        vector<ll> W(M);
        for (int j = 0; j < M; ++j) {
            cin >> U[j] >> V[j] >> W[j];
        }
        for (int j = 0; j < Q; ++j) {
            cin >> A[j] >> B[j];
        }
        solve(N, M, Q, U, V, W, A, B);
    }
}